File: packages/telegram-control/python/telegram_control/alerts.py
"""Alert formatting utilities for trading notifications."""

from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import List, Tuple


class AlertLevel(Enum):
    """Alert severity levels."""
    
    INFO = "ℹ️"
    SUCCESS = "✅"
    WARNING = "⚠️"
    ERROR = "❌"
    CRITICAL = "🚨"


@dataclass
class AlertBuilder:
    """Builder for trading alerts."""
    
    level: AlertLevel
    title: str
    fields: List[Tuple[str, str]] = field(default_factory=list)
    timestamp: datetime = field(default_factory=datetime.utcnow)
    
    @classmethod
    def info(cls, title: str) -> "AlertBuilder":
        """Create an info alert."""
        return cls(level=AlertLevel.INFO, title=title)
    
    @classmethod
    def success(cls, title: str) -> "AlertBuilder":
        """Create a success alert."""
        return cls(level=AlertLevel.SUCCESS, title=title)
    
    @classmethod
    def warning(cls, title: str) -> "AlertBuilder":
        """Create a warning alert."""
        return cls(level=AlertLevel.WARNING, title=title)
    
    @classmethod
    def error(cls, title: str) -> "AlertBuilder":
        """Create an error alert."""
        return cls(level=AlertLevel.ERROR, title=title)
    
    @classmethod
    def critical(cls, title: str) -> "AlertBuilder":
        """Create a critical alert."""
        return cls(level=AlertLevel.CRITICAL, title=title)
    
    def field(self, name: str, value: str) -> "AlertBuilder":
        """Add a field to the alert."""
        self.fields.append((name, value))
        return self
    
    def price(self, name: str, value: float, currency: str = "USD") -> "AlertBuilder":
        """Add a price field with formatting."""
        return self.field(name, format_price(value, currency))
    
    def percentage(self, name: str, value: float) -> "AlertBuilder":
        """Add a percentage field."""
        emoji = "📈" if value >= 0 else "📉"
        return self.field(name, f"{emoji} {value:.2f}%")
    
    def build(self) -> str:
        """Build the alert message."""
        lines = [f"{self.level.value} **{self.title}**"]
        
        for name, value in self.fields:
            lines.append(f"• *{name}*: {escape_markdown(value)}")
        
        return "\n".join(lines)
    
    def __str__(self) -> str:
        """String representation."""
        return self.build()


def format_price(value: float, currency: str = "USD") -> str:
    """Format a price value."""
    currency = currency.upper()
    
    if currency in ("USD", "USDC", "USDT"):
        return f"${value:.2f}"
    elif currency == "BTC":
        return f"₿{value:.8f}"
    elif currency == "ETH":
        return f"Ξ{value:.6f}"
    else:
        return f"{value:.4f} {currency}"


def format_percentage(value: float) -> str:
    """Format a percentage change."""
    sign = "+" if value >= 0 else ""
    return f"{sign}{value:.2f}%"


def escape_markdown(text: str) -> str:
    """Escape markdown special characters."""
    chars = ["_", "*", "[", "]", "(", ")", "~", "`", ">", "#", "+", "-", "=", "|", "{", "}", ".", "!"]
    for char in chars:
        text = text.replace(char, f"\\{char}")
    return text


# Pre-built alert templates
def trade_executed(
    side: str,
    symbol: str,
    amount: float,
    price: float,
    total: float
) -> str:
    """Alert for executed trade."""
    return (
        AlertBuilder.success("Trade Executed")
        .field("Side", side)
        .field("Symbol", symbol)
        .field("Amount", str(amount))
        .price("Price", price)
        .price("Total", total)
        .build()
    )


def price_alert(symbol: str, price: float, change_pct: float, threshold: float) -> str:
    """Price movement alert."""
    if abs(change_pct) >= threshold * 2:
        level = AlertLevel.CRITICAL
    elif abs(change_pct) >= threshold:
        level = AlertLevel.WARNING
    else:
        level = AlertLevel.INFO
    
    return (
        AlertBuilder(level, f"{symbol} Price Alert")
        .price("Current Price", price)
        .percentage("Change", change_pct)
        .build()
    )


def risk_alert(metric: str, value: float, threshold: float) -> str:
    """Risk threshold alert."""
    return (
        AlertBuilder.warning("Risk Threshold Reached")
        .field("Metric", metric)
        .field("Current Value", f"{value:.2f}")
        .field("Threshold", f"{threshold:.2f}")
        .build()
    )

File: packages/telegram-control/rust/src/alerts.rs
use chrono::{DateTime, Utc};
use std::fmt::Write;

/// Alert severity levels
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum AlertLevel {
    Info,
    Success,
    Warning,
    Error,
    Critical,
}

impl AlertLevel {
    pub fn emoji(&self) -> &'static str {
        match self {
            AlertLevel::Info => "ℹ️",
            AlertLevel::Success => "✅",
            AlertLevel::Warning => "⚠️",
            AlertLevel::Error => "❌",
            AlertLevel::Critical => "🚨",
        }
    }
}

/// Builder for trading alerts
#[derive(Clone, Debug)]
pub struct AlertBuilder {
    level: AlertLevel,
    title: String,
    fields: Vec<(String, String)>,
    timestamp: DateTime<Utc>,
}

impl AlertBuilder {
    pub fn new(level: AlertLevel, title: impl Into<String>) -> Self {
        Self {
            level,
            title: title.into(),
            fields: Vec::new(),
            timestamp: Utc::now(),
        }
    }
    
    pub fn info(title: impl Into<String>) -> Self {
        Self::new(AlertLevel::Info, title)
    }
    
    pub fn success(title: impl Into<String>) -> Self {
        Self::new(AlertLevel::Success, title)
    }
    
    pub fn warning(title: impl Into<String>) -> Self {
        Self::new(AlertLevel::Warning, title)
    }
    
    pub fn error(title: impl Into<String>) -> Self {
        Self::new(AlertLevel::Error, title)
    }
    
    pub fn critical(title: impl Into<String>) -> Self {
        Self::new(AlertLevel::Critical, title)
    }
    
    /// Add a field to the alert
    pub fn field(mut self, name: impl Into<String>, value: impl Into<String>) -> Self {
        self.fields.push((name.into(), value.into()));
        self
    }
    
    /// Add a price field with formatting
    pub fn price(self, name: impl Into<String>, value: f64, currency: &str) -> Self {
        self.field(name, format_price(value, currency))
    }
    
    /// Add a percentage field
    pub fn percentage(self, name: impl Into<String>, value: f64) -> Self {
        let emoji = if value >= 0.0 { "📈" } else { "📉" };
        self.field(name, format!("{} {:.2}%", emoji, value))
    }
    
    /// Build the alert message
    pub fn build(self) -> String {
        let mut msg = String::new();
        writeln!(&mut msg, "{} **{}**", self.level.emoji(), self.title).unwrap();
        
        for (name, value) in self.fields {
            writeln!(&mut msg, "• *{}*: {}", name, escape_markdown(&value)).unwrap();
        }
        
        msg
    }
}

/// Format a price value
pub fn format_price(value: f64, currency: &str) -> String {
    match currency.to_uppercase().as_str() {
        "USD" | "USDC" | "USDT" => format!("${:.2}", value),
        "BTC" => format!("₿{:.8}", value),
        "ETH" => format!("Ξ{:.6}", value),
        _ => format!("{:.4} {}", value, currency),
    }
}

/// Format a percentage change
pub fn format_percentage(value: f64) -> String {
    let sign = if value >= 0.0 { "+" } else { "" };
    format!("{}{:.2}%", sign, value)
}

/// Escape markdown special characters
fn escape_markdown(text: &str) -> String {
    text.replace('_', "\\_")
        .replace('*', "\\*")
        .replace('[', "\\[")
        .replace(']', "\\]")
        .replace('(', "\\(")
        .replace(')', "\\)")
        .replace('~', "\\~")
        .replace('`', "\\`")
        .replace('>', "\\>")
        .replace('#', "\\#")
        .replace('+', "\\+")
        .replace('-', "\\-")
        .replace('=', "\\=")
        .replace('|', "\\|")
        .replace('{', "\\{")
        .replace('}', "\\}")
        .replace('.', "\\.")
        .replace('!', "\\!")
}

/// Pre-built alert templates for trading
pub mod templates {
    use super::*;
    
    pub fn trade_executed(
        side: &str,
        symbol: &str,
        amount: f64,
        price: f64,
        total: f64,
    ) -> String {
        AlertBuilder::success("Trade Executed")
            .field("Side", side)
            .field("Symbol", symbol)
            .price("Amount", amount, "")
            .price("Price", price, "USD")
            .price("Total", total, "USD")
            .build()
    }
    
    pub fn price_alert(symbol: &str, price: f64, change_pct: f64, threshold: f64) -> String {
        let level = if change_pct.abs() >= threshold * 2.0 {
            AlertLevel::Critical
        } else if change_pct.abs() >= threshold {
            AlertLevel::Warning
        } else {
            AlertLevel::Info
        };
        
        AlertBuilder::new(level, &format!("{} Price Alert", symbol))
            .price("Current Price", price, "USD")
            .percentage("Change", change_pct)
            .build()
    }
    
    pub fn risk_alert(metric: &str, value: f64, threshold: f64) -> String {
        AlertBuilder::warning("Risk Threshold Reached")
            .field("Metric", metric)
            .field("Current Value", format!("{:.2}", value))
            .field("Threshold", format!("{:.2}", threshold))
            .build()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_alert_builder() {
        let alert = AlertBuilder::success("Test Alert")
            .field("Key", "Value")
            .build();
        
        assert!(alert.contains("✅ **Test Alert**"));
        assert!(alert.contains("Key"));
        assert!(alert.contains("Value"));
    }
    
    #[test]
    fn test_format_price() {
        assert_eq!(format_price(100.5, "USD"), "$100.50");
        assert_eq!(format_price(0.00123456, "BTC"), "₿0.00123456");
    }
}
